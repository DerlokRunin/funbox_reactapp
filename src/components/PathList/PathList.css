.pathlist__ul{
    padding: 0px;
}
.pathlist__ul li{
    list-style: none;
}
.placeholder {
    background: rgb(251, 245, 228);
    padding: 10px;
    margin-top: 10px;
}
.placeholder:before {
    content: "Переместить сюда";
    color: rgb(208, 208, 208);
}