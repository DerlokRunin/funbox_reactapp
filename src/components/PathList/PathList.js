import React, {Component} from 'react';
import PropTypes from "prop-types";
import './PathList.css';
import Path from './Path/Path';

class PathList extends Component{

    static propTypes = {
        points: PropTypes.array,//точки с координатами
        onDelete: PropTypes.func,//действия при удалении
        onRevertPoints: PropTypes.func,//действия при перемещении
    };

    constructor(props){
        super(props);


        //Пустышка drag and drop
        this.placeholder = document.createElement("li");
        this.placeholder.className = "placeholder";
    }

    render(){

        const {points, onDelete} = this.props;
        const pathElements = points.map((point, index) =>
            <li
                dataset_id={index}
                key={point.id}
                draggable='true'
                onDragEnd={this.dragEnd}
                onDragStart={this.dragStart}
            >
                <Path point={point} onDelete={onDelete.bind(this, point.id)}/>
            </li>
        );

        return (
            <div>
                <ul className="pathlist__ul" onDragOver={this.dragOver}>
                    {pathElements}
                </ul>
            </div>
        )
    }

    //начало перемещения
    dragStart = (event) => {
        this.dragged = event.currentTarget;
        event.dataTransfer.effectAllowed = 'move';
        event.dataTransfer.setData('text/html', this.dragged);
    };
    //окончание перемещения
    dragEnd = (event) => {
        this.dragged.style.display = 'block';
        try {
            this.dragged.parentNode.removeChild(this.placeholder);
        }catch(e){}

        // меняем местами данные
        const {onRevertPoints} = this.props;
        var from = Number(this.dragged.getAttribute('dataset_id'));
        var to = Number(this.over.getAttribute('dataset_id'));
        onRevertPoints(to, from);
    };
    //наведение
    dragOver = (event) => {
        event.preventDefault();
        this.dragged.style.display = "none";
        if(event.target.className === 'path__point') {
            this.over = event.target.parentNode.parentNode;
            this.before = !this.before;
            if(this.before)
                this.over.parentNode.insertBefore(this.placeholder, this.over);
            else
                this.over.parentNode.insertBefore(this.placeholder, this.over.nextSibling);
        }else
        if(event.target.className === 'placeholder') return;


    };
}

export default PathList;