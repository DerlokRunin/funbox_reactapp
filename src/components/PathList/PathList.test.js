import React from 'react';
import PathList from './PathList';


it('проверка рендеринга списка', () => {
    let points = [
        {id:1, title:'test', coordinates: [42.423, 35.353] },
        {id:2, title:'test2', coordinates: [42.423, 35.353] },
        {id:3, title:'test3', coordinates: [42.423, 35.353] }
    ];
    const wrapper = shallow(
        <PathList points={points} onDelete={function () {}} onRevertPoints={function () {}}/>
    );
    expect(wrapper).toMatchSnapshot();
});