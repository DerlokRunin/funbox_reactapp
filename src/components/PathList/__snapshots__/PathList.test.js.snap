// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`проверка рендеринга списка 1`] = `
<div>
  <ul
    className="pathlist__ul"
    onDragOver={[Function]}
  >
    <li
      dataset_id={0}
      draggable="true"
      key="1"
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <Path
        onDelete={[Function]}
        point={
          Object {
            "coordinates": Array [
              42.423,
              35.353,
            ],
            "id": 1,
            "title": "test",
          }
        }
      />
    </li>
    <li
      dataset_id={1}
      draggable="true"
      key="2"
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <Path
        onDelete={[Function]}
        point={
          Object {
            "coordinates": Array [
              42.423,
              35.353,
            ],
            "id": 2,
            "title": "test2",
          }
        }
      />
    </li>
    <li
      dataset_id={2}
      draggable="true"
      key="3"
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <Path
        onDelete={[Function]}
        point={
          Object {
            "coordinates": Array [
              42.423,
              35.353,
            ],
            "id": 3,
            "title": "test3",
          }
        }
      />
    </li>
  </ul>
</div>
`;
